### ISSUE TITLE
[Bug Report]:Unable to produce tombstone records

### ISSUE THREAD

**Issue by adimoraret** [Farfetch/kafkaflow#516]
### Prerequisites

- [X] I have searched [issues](https://github.com/Farfetch/kafkaflow/issues) to ensure it has not already been reported

### Description

I am unable to produce tombstone records. It looks like the message value is sent as byte[0] instead of null.

### Steps to reproduce

```csharp
services.AddKafka(kafka => kafka
            .AddCluster(cluster => cluster
                .WithBrokers(new[] { kafkaBrokers })
                .AddProducer("test", producer => producer
                    .AddMiddlewares(middlewares => middlewares
                        .AddSerializer<JsonCoreSerializer>(_ => new JsonCoreSerializer(options))
                    )
                    .WithAcks(Acks.All)
                )
            )
        );

await producer.ProduceAsync("test", Guid.NewGuid().ToString(), null);
```

### Expected behavior

Message value should be sent as null

### Actual behavior

Message value is sent as byte[0]

### KafkaFlow version

3.0.3

**Comment by tomaszprasolek** [comment]
Do you have any error while producing the message?

**Comment by adimoraret** [comment]
No, there is no error.  But it won't produce tombstone records when using a kafka upsert source: https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/upsert-kafka/ .

**Comment by JoaoRodriguesGithub** [comment]
Like we discussed on kafkaflow on [slack](https://kafkaflow.slack.com/archives/C0178T4GC04/p1706562993854229), by using a custom serializer, you can bypass this limitation by checking if the message is null before serializing the message value.

This should procude a message where the message value is an empty byte[] witch is the thumbstone.

Meanwhile, we will address this issue, by creating a pull request to fix the JsonCoreSerializer.

**Comment by adrianmoraret-mdsol** [comment]
I have tried that approach, end to end, using kafka-upsert source and iceberg table destination in AWS. It sends indeed an empty byte array, but that did not translate into a tombstone record. 
If I read correctly this https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/upsert-kafka/, it would only work if we send null for message. 
Maybe I missed something, but I'm happy to test your PR once ready.

**Comment by JoaoRodriguesGithub** [comment]
Fixing the issue with JsonCoreSerializer when producing a message with a null value it results in a tombstone record where the message value is an empty byte[].

> "A null payload is a payload with 0 bytes" 
[reference here](https://www.linkedin.com/pulse/tombstone-record-kafka-under-hood-sumon-mal)

```csharp
public Task SerializeAsync(object? message, Stream output, ISerializerContext context)
    {
        if (message == null)
        {
            return Task.CompletedTask;
        }

        return SerializeNonNullMessageAsync(message, output);
    }
```

As an alternative, you can always use the native Kafka client to confirm that it also produces a tombstone record with an empty byte[] like KafkaFlow.

**Comment by esskar** [comment]
> Fixing the issue with JsonCoreSerializer when producing a message with a null value it results in a tombstone record where the message value is an empty byte[].
> 
> > "A null payload is a payload with 0 bytes"
> > [reference here](https://www.linkedin.com/pulse/tombstone-record-kafka-under-hood-sumon-mal)
> 
> ```cs
> public Task SerializeAsync(object? message, Stream output, ISerializerContext context)
>     {
>         if (message == null)
>         {
>             return Task.CompletedTask;
>         }
> 
>         return SerializeNonNullMessageAsync(message, output);
>     }
> ```
> 
> As an alternative, you can always use the native Kafka client to confirm that it also produces a tombstone record with an empty byte[] like KafkaFlow.

I resolved it the sameway for the ProtobufSerializer. Is there an idea to fix it in general?

**Comment by mrt181** [comment]
I don't think the fix works, see my comment in the PR
[#547](https://github.com/Farfetch/kafkaflow/pull/547#issuecomment-2293271120)

**Comment by mrt181** [comment]
for anyone that struggles with tombstones, this is what I had to do to properly create a tombstone that is also recognized as such for example in ksqlDB.

```c#
.AddProducer<Confluent.Kafka.Null>(producer => { })
```

and then use this producer when sending a `null` message with a key to a topic.

**Comment by epbensimpson** [comment]
I don't believe that this issue is fixed by #547 and needs to be reopened. With Kafkaflow, producing a message with a `null` value results in an empty `byte[]` being sent to Kafka. Per librdkafka [docs](https://github.com/confluentinc/librdkafka/wiki/FAQ#how-are-tombstones-deletes-represented-in-compacted-topics), the value must be `null` to act as a tombstone. Also for non-compacted topics this means you cannot differentiate between and empty message and a `null`. 

If I use `Confluent.Kafka` directly I can successfully produce a message with an actually `null` value:

```csharp
[Fact]
public async Task EmptyOrNullValueTest()
{
    await _producer.ProduceAsync(_fixture.ScheduledTopicName,
        new Message<string, byte[]> { Key = "foo", Value = null! });

    await _producer.Producer.ProduceAsync(_fixture.ScheduledTopicName,
        new Message<string, byte[]> { Key = "foo", Value = [] });

    var nullResult = _consumer.Consume(TimeSpan.FromSeconds(10));
    var emptyResult = _consumer.Consume(TimeSpan.FromSeconds(10));

    nullResult.Message.Value.Should().BeNull();
    emptyResult.Message.Value.Should().BeEmpty();
}
```

I don't understand the outcome of [this conversation](https://github.com/Farfetch/kafkaflow/pull/547#discussion_r1538186873) around whether this should be `null` or an empty array in the PR. It _must_ be `null`! 

The PR did basically nothing because the change results in behavior that is.... exactly how OP described the issue:

> I am unable to produce tombstone records. It looks like the message value is sent as byte[0] instead of null.

**Comment by AlexeyRaga** [comment]
I believe this issue needs to be re-open, as per @epbensimpson comment.

Currently KafkaFlow contradicts Kafka guidelines, `librdkafka` and `Confluent` libraries for .NET.

It also makes it impossible to have compacted topics with deletion, since it is currently not possible to send `null` for deleting records for a specified key.

Which, in turn, makes it impossible to use Kafka in scenarios when GDPR and other compliance rules might be in play.

It also seems to be a simple one-liner change here:
https://github.com/Farfetch/kafkaflow/blob/master/src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs#L42

I am happy to make a PR, can we agree on it and re-open this issue?

**Comment by MiguelCosta** [comment]
Thank you very much for all the information and questions you have shared 🙂

This is indeed an important point that we need to review. Our team will organize and work on resolving this issue. 
We will provide feedback as soon as possible.

**Comment by Matthew-Davey** [comment]
Is the problem in fact [here](https://github.com/Farfetch/kafkaflow/blob/bf94613dfc641431c42a91e01e1bd12ce782ca0a/src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs#L42-L59)?

The serializer is bypassed if `context.Message.Value` is null, however on line [#59](https://github.com/Farfetch/kafkaflow/blob/bf94613dfc641431c42a91e01e1bd12ce782ca0a/src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs#L59) the message value switched from `null` to `Array.Empty<byte>()` anyway.

**Comment by esskar** [comment]
So i see that #630 is merged in 4.0, cant we have a bugfix before for 3.X?
